src/Extensions/FakeVimProxy.cpp

```diff
@@ -0,0 +1,506 @@
+
+void FakeVimProxy::checkForElectricCharacter(bool *result, QChar c)
+{
+    *result = c == '{' || c == '}';
+}
+
+int FakeVimProxy::firstNonSpace(const QString &text)
+{
+    int indent = 0;
+    while (indent < text.length() && text.at(indent) == ' ')
+        ++indent;
+    return indent;
+}
+
+void FakeVimProxy::updateExtraSelections()
+{
+    QTextEdit *editor = qobject_cast<QTextEdit *>(m_widget);
+    QPlainTextEdit *plainEditor = qobject_cast<QPlainTextEdit *>(m_widget);
+    if (editor)
+    {
+        editor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+    else if (plainEditor)
+    {
+        plainEditor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+}
+
+bool FakeVimProxy::wantSaveAndQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.cmd == "wq";
+}
+
+bool FakeVimProxy::wantSave(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("w", "write") || cmd.matches("wa", "wall");
+}
+
+bool FakeVimProxy::wantQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("q", "quit") || cmd.matches("qa", "qall");
+}
+
+bool FakeVimProxy::save()
+{
+    if (!hasChanges())
+        return true;
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        return window->save(true, "Vim Save");
+
+    return false;
+}
+
+void FakeVimProxy::cancel()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+    {
+        if (hasChanges())
+        {
+            if (window->closeConfirm())
+                window->closeWindow();
+        }
+        else
+            invalidate();
+    }
+}
+void FakeVimProxy::invalidate()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        window->closeWindow();
+}
+
+bool FakeVimProxy::hasChanges()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        return window->isTextChanged();
+    return false;
+}
+
+QTextDocument *FakeVimProxy::document() const
+{
+    QTextDocument *doc = NULL;
+    if (QPlainTextEdit *ed = qobject_cast<QPlainTextEdit *>(m_widget))
+        doc = ed->document();
+    else if (QTextEdit *ed = qobject_cast<QTextEdit *>(m_widget))
+        doc = ed->document();
+    return doc;
+}
+
+QString FakeVimProxy::content() const
+{
+    return document()->toPlainText();
+}
+
+void FakeVimProxy::initHandler(FakeVim::Internal::FakeVimHandler *handler)
+{
+    handler->handleCommand(_("set nopasskeys"));
+    handler->handleCommand(_("set nopasscontrolkey"));
+
+    // Set some Vim options.
+    handler->handleCommand(_("set expandtab"));
+    handler->handleCommand(_("set shiftwidth=8"));
+    handler->handleCommand(_("set tabstop=%1").arg(SettingsHelper::getTabWidth()));
+    handler->handleCommand(_("set autoindent"));
+    handler->handleCommand(_("set smartindent"));
+
+    handler->handleCommand("source " + SettingsHelper::getFakeVimRCLocation());
```

Comment:
Check the location exists, or is not empty first.

Reply:
Thanks, forgot it!

Reply:
I am thinking that instead of loading it from File We can provide a `QPlainTextEdit` where user puts their config. Like Clang Format Config. The benefit would be that it will be more portable, (configuration will be exported along with rest of settings). It is because usually FakeVim does not supports full Vim Commands, plus in most of users vim configs the configs for VimPlug/Pathogen are present which is completely unsupported in FakeVim.